File: packages/@forma/ts/src/Env.ts
import type { KValue } from "./evaluator/types.js";

/**
 * Immutable lexical scope chain.
 *
 * `bindMutable` creates a slot that can be updated after creation — used by
 * `define` so a function's closure can reference itself for recursion.
 */
export class Env {
  private readonly bindings: Map<string, KValue>;
  private readonly parent: Env | null;

  private constructor(bindings: Map<string, KValue>, parent: Env | null) {
    this.bindings = bindings;
    this.parent = parent;
  }

  static empty(): Env {
    return new Env(new Map(), null);
  }

  static from(record: Record<string, KValue>): Env {
    return new Env(new Map(Object.entries(record)), null);
  }

  lookup(name: string): KValue | undefined {
    const val = this.bindings.get(name);
    if (val !== undefined) return val;
    // Check for explicit null binding
    if (this.bindings.has(name)) return null;
    if (this.parent) return this.parent.lookup(name);
    return undefined;
  }

  has(name: string): boolean {
    if (this.bindings.has(name)) return true;
    if (this.parent) return this.parent.has(name);
    return false;
  }

  bindingNames(): readonly string[] {
    const names = this.parent ? [...this.parent.bindingNames()] : [];
    for (const name of this.bindings.keys()) {
      if (!names.includes(name)) names.push(name);
    }
    return names;
  }

  bind(name: string, value: KValue): Env {
    const newBindings = new Map(this.bindings);
    newBindings.set(name, value);
    return new Env(newBindings, this.parent);
  }

  /**
   * Create a child env with a mutable slot. Returns the new env and a setter
   * that mutates the slot in place — used for recursive `define` bindings.
   */
  bindMutable(name: string, initial: KValue): { env: Env; set: (value: KValue) => void } {
    const childMap = new Map<string, KValue>();
    childMap.set(name, initial);
    const childEnv = new Env(childMap, this);
    return {
      env: childEnv,
      set: (value: KValue) => {
        childMap.set(name, value);
      },
    };
  }

  extend(bindings: Record<string, KValue>): Env {
    return new Env(new Map(Object.entries(bindings)), this);
  }

  /**
   * Return a copy of this env with a new parent. Used to chain prelude
   * macros beneath a user-provided environment.
   */
  withParent(parent: Env): Env {
    return new Env(new Map(this.bindings), parent);
  }
}

File: packages/@forma/ocaml/lib/env.ml
module String_map = Map.Make (String)

type t = { bindings : (string * Value.t) list; index : Value.t String_map.t }

let empty = { bindings = []; index = String_map.empty }
let lookup name env = String_map.find_opt name env.index

let bind name value env =
  {
    bindings = (name, value) :: env.bindings;
    index = String_map.add name value env.index;
  }

let extend bindings env =
  {
    bindings = bindings @ env.bindings;
    index =
      List.fold_right
        (fun (name, value) index -> String_map.add name value index)
        bindings env.index;
  }

let bindings env = env.bindings
let of_bindings bindings = extend bindings empty
let length env = List.length env.bindings

let remove_names names env =
  env.bindings
  |> List.filter (fun (name, _) -> not (List.mem name names))
  |> of_bindings
